File: python/conv_nhwc.py
from torch import nn
import torch
import numpy as np

def write_binary_file(file_path, array):
    with open(file_path, 'wb') as f:
        array.tofile(f)

def write_npy_file(file_path, array):
    np.save(file_path, array)

INPUT_CHANNELS = 64
OUTPUT_CHANNELS = 128
KERNEL_SIZE = 3
STRIDE = 1
PADDING = 0
HEIGHT = 128
WIDTH = 128

if __name__ == "__main__":

    conv = nn.Conv2d(
        OUTPUT_CHANNELS,
        INPUT_CHANNELS,
        kernel_size=(KERNEL_SIZE, KERNEL_SIZE),
        stride=STRIDE,
        padding=PADDING,
    )

    temp = 0
    input_matrix = []
    for d in range(HEIGHT):
        depth_slice = []
        for i in range(WIDTH):
            row = []
            temp = 0
            for j in range(INPUT_CHANNELS):
                row.append(temp)
                temp += 1
            depth_slice.append(row)
        input_matrix.append(depth_slice)
    
    # Reshape to NHWC format
    input_tensor = torch.Tensor(input_matrix).reshape(1, HEIGHT, WIDTH, INPUT_CHANNELS)
    print("Input shape (NHWC):", input_tensor.shape)
    
    # Permute to NCHW format
    input_tensor = input_tensor.permute(0, 3, 1, 2)
    print("Permuted input shape (NCHW):", input_tensor.shape)

    filters_matrix = []
    temp = 0
    for k in range(KERNEL_SIZE):
        kernel_matrix = []
        for d in range(KERNEL_SIZE):
            depth_slice = []
            for i in range(INPUT_CHANNELS):
                row = []
                temp = 0
                for j in range(OUTPUT_CHANNELS):
                    row.append(temp)
                    temp += 1
                depth_slice.append(row)
            kernel_matrix.append(depth_slice)
        filters_matrix.append(kernel_matrix)

    # Reshape to KKIO format
    filters_tensor = torch.Tensor(filters_matrix).reshape(
        KERNEL_SIZE, KERNEL_SIZE, INPUT_CHANNELS, OUTPUT_CHANNELS
    )
    print("Weights shape (KKIO):", filters_tensor.shape)
    
    # Permute to OIKK format
    filters_tensor = filters_tensor.permute(3, 2, 0, 1)
    print("Permuted weights shape (OIKK):", filters_tensor.shape)

    # Export input and kernel tensors to numpy files
    write_npy_file('input_py.npy', input_tensor.numpy())
    write_npy_file('kernel_py.npy', filters_tensor.numpy())

    # Re-assign the permuted tensors to the convolution layer
    conv.weight.data = filters_tensor
    conv.bias.data = torch.zeros(OUTPUT_CHANNELS)

    # Perform convolution
    output = conv(input_tensor)
    print("Output shape:", output.shape)

    # Permute output from NCHW to NHWC
    output = output.permute(0, 2, 3, 1)
    print("Permuted output shape (NHWC):", output.shape)

    # Export the output
    output_np = output.detach().numpy()
    write_npy_file('output_py.npy', output_np)
    output_1d = output_np.flatten()    # Flatten to 1D array for binary file
    write_binary_file('py_nhwc_output.bin', output_1d)

    print("3D Convolution complete. Output written to 'py_nhwc_output.bin'.")

File: cpp/conv_nhwc.cpp
#include <iostream>
#include <vector>
#include <fstream>

#define INPUT_HEIGHT 128
#define INPUT_WIDTH 128
#define INPUT_CHANNEL 64
#define KERNEL_HEIGHT 3
#define KERNEL_WIDTH 3
#define KERNEL_CHANNEL 64
#define OUTPUT_CHANNEL 128

using namespace std;

void write_to_binary(const string &filename, const vector<float> &data)
{
    ofstream file(filename, ios::binary);
    file.write(reinterpret_cast<const char *>(data.data()), data.size() * sizeof(float));
    file.close();
}

int main()
{
    vector<vector<vector<vector<float>>>> input;
    vector<vector<vector<vector<float>>>> kernel;

    int temp = 0;

    vector<vector<vector<float>>> channel;
    for (int i = 0; i < INPUT_HEIGHT; i++)
    {
        vector<vector<float>> col;
        for (int j = 0; j < INPUT_WIDTH; j++)
        {
            temp = 0;
            vector<float> row;
            for (int k = 0; k < INPUT_CHANNEL; k++)
            {
                row.push_back(temp);
                temp++;
            }
            col.push_back(row);
        }
        channel.push_back(col);
    }
    input.push_back(channel);

    int N = input.size();
    int H = input[0].size();
    int W = input[0][0].size();
    int C = input[0][0][0].size();
    cout << "Input: \n";
    cout << "N:" << N << " H:" << H << " W:" << W << " C:" << C << "\n";

    temp = 0;

    for (int c = 0; c < KERNEL_HEIGHT; c++)
    {
        vector<vector<vector<float>>> channel;
        for (int i = 0; i < KERNEL_WIDTH; i++)
        {
            vector<vector<float>> col;
            for (int j = 0; j < KERNEL_CHANNEL; j++)
            {
                temp=0;
                vector<float> row;
                for (int k = 0; k < OUTPUT_CHANNEL; k++)
                {
                    row.push_back(temp);
                    temp++;
                }
                col.push_back(row);
            }
            channel.push_back(col);
        }
        kernel.push_back(channel);
    }

    H = kernel.size();
    W = kernel[0].size();
    C = kernel[0][0].size();
    int OC = kernel[0][0][0].size();
    cout << "Kernel: \n";
    cout << "H:" << H << " W:" << W << " C:" << C << " OC:" << OC << "\n";

    int stride = 1;
    int padding = 0;

    int output_height = ((INPUT_HEIGHT - KERNEL_HEIGHT + (2 * padding)) / stride) + 1;
    int output_width = ((INPUT_WIDTH - KERNEL_WIDTH + (2 * padding)) / stride) + 1;

    cout << "Output Height: " << output_height << endl;
    cout << "Output Width: " << output_width << endl;

    vector<vector<vector<vector<float>>>> output(1, vector<vector<vector<float>>>(output_height, vector<vector<float>>(output_width, vector<float>(OUTPUT_CHANNEL, 0))));

    for (int ih = 0; ih < output_height; ih++)
    {
        for (int iw = 0; iw < output_width; iw++)
        {
            for (int outc = 0; outc < OUTPUT_CHANNEL; outc++)
            {
                float sum = 0;
                for (int kh = 0; kh < KERNEL_HEIGHT; kh++)
                {
                    for (int kw = 0; kw < KERNEL_WIDTH; kw++)
                    {
                        for (int kc = 0; kc < KERNEL_CHANNEL; kc++)
                        {
                            int oh = ih * stride + kh;
                            int ow = iw * stride + kw;
                            sum += input[0][oh][ow][kc] * kernel[kh][kw][kc][outc]; 
                        }
                    }
                }
                output[0][ih][iw][outc] = sum;
            }
        }
    }

    vector<float> output_1d;
    for (int i = 0; i < output_height; i++)
    {
        for (int j = 0; j < output_width; j++)
        {
            for (int k = 0; k < OUTPUT_CHANNEL; k++)
            {
                output_1d.push_back(output[0][i][j][k]);
            }
        }
    }
    
    write_to_binary("cpp_nhwc_output.bin", output_1d);

    cout << "3D Convolution complete. Output written to 'cpp_nhwc_output.bin'." << endl;

    return 0;
}
